MatrixFree: enable DiagonalMatrix in loops

bergbauer opened the pull request:
@kronbichler

masterleinad reviewed (commented):
What about a test? What's the previous behavior?

bergbauer commented:
> What about a test? What's the previous behavior?

You got a compile error when you tried to use `loop`. I will add a test 👍🏽

kronbichler commented:
If I understand correctly, I do not really the fact that we need to specialize for a non-vector type `DiagonalMatrix<VectorType>` in a templated loop. If I understand your description correctly, you want to circumvent a compile error? Compilation should be fine, but I don't see how we could realistically cover all cases. I would like to zero the vector for vector types, and in all other cases simply do nothing and make it the user's responsibility. We could add an assertion that makes sure that if a user requests zeroing a vector, but we do not know how to do that, say that we can't zero. What do you think?

kronbichler reviewed (approved):
Thanks a lot, this is really nice and clearly assigns the responsibility to the user!

kronbichler commented:
/rebuild

bergbauer commented:
I changed the type trait to check for the assignment operator. A remaining question is whether we want to add `operator=(const Number)` to `DiagonalMatrix`.

The off-diagonal block instantiation in `AffineConstraints` was missing for `DiagonalMatrix<Vector>` that's why I added it.
